### PR TITLE
allclose_units()

### PR DESCRIPTION
**Pull request by berceanu**
Address https://github.com/yt-project/unyt/issues/92

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Hey sorry for taking a few days to respond to this.

So this does look good, although it looks like there's some lines you've added that don't have test coverage. New code in unyt needs to have 100% test coverage. In particular it looks like the error branches in `allclose_units` aren't being tested at the moment.

In addition, making `unyt.testing` public is probably a good change, but more work needs to happen to bring the code up to the standard of the other public modules in unyt. In particular, the `process_warning` function probably shouldn't be public, so it should be renamed to `_process_warning`. In addition, you'll need to make sure that both `unyt.testing` and `assert_allclose_units` shows up in the API documentation. I would also make it so that you can do `from unyt import assert_allclose_units` by reimporting `assert_allclose_units` in the top-level `unyt` namespace. Both of these can be accomplished by editing `unyt/__init__.py`, the docstrings of this file are rendered in the API docs page (e.g. https://unyt.readthedocs.io/en/stable/modules/unyt.html), and this file also defines the top-level `unyt` namespace.

To make sure that `unyt.testing` shows up in the built docs you'll need to build the docs locally and double-check that's the case. To build the docs go into the `docs` directory in the repo and do `make html`. You'll need to have sphinx installed for that to work correctly.

It would also probably be worthwhile to mention `assert_allclose_units` in the narrative documentation, particularly the section on using `unyt` as a part of a larger python codebase.

File: setup.cfg [hunk 1]
```diff
@@ -36,7 +36,7 @@ omit =
     versioneer.py
     unyt/_version.py
     unyt/_on_demand_imports.py
-    unyt/_testing.py
```
Comment (ngoldbaum) [inline setup.cfg]: If we're exposing this publicly I don't think we can get away with excluding it from test coverage anymore.

**Comment by ngoldbaum** [comment]
btw if you're not up for making `unyt.testing` public that's cool, just make it private again and open an issue about maybe making it public in the future.

**Comment by berceanu** [comment]
I’ll give it a shot :)

File: setup.cfg (hunk 1 above)
Comment (berceanu) [reply setup.cfg]: Added `tests/test_unyt_testing.py`.

**Comment by berceanu** [comment]
Done. Care to have another look?

File: unyt/array.py [hunk 2]
```diff
@@ -2280,3 +2281,37 @@ def savetxt(
...
+
+    if not isinstance(atol, unyt_array):
+        at = unyt_quantity(atol, des.units)
+    else:
+        at = atol
+
+    try:
+        at = at.in_units(act.units)
+    except UnitOperationError:
+        return False
```
Comment (ngoldbaum) [inline unyt/array.py]: codecov reports that this line is still untested

File: unyt/tests/test_unyt_testing.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    a1 = unyt_array([1.0, 2.0, 3.0], "cm")
+    a2 = unyt_array([1.0, 2.0, 3.0], "cm")
+    with pytest.raises(UnitConversionError):
+        assert_allclose_units(a1, a2, atol=unyt_quantity(0.0, "kg"))
+
+
+# from unyt import assert_allclose_units, m
+#  actual = [1e-5, 1e-3, 1e-1] * m
+#  desired = actual.to("cm")
+#  assert_allclose_units(actual, desired)
```
[the remark this answers is not in the document]
Comment (berceanu) [reply unyt/tests/test_unyt_testing.py]: Sorry! Done.

File: unyt/array.py (hunk 2 above)
Comment (berceanu) [reply unyt/array.py]: Yes, I saw that and I was wondering how to hit that exception..

File: unyt/__init__.py [hunk 4]
```diff
@@ -89,6 +90,7 @@
 from unyt.unit_object import Unit, define_unit  # NOQA: F401
 from unyt.unit_registry import UnitRegistry  # NOQA: F401
 from unyt.unit_systems import UnitSystem  # NOQA: F401
+from unyt.testing import assert_allclose_units  # NOQA: F401
```
[the remark this answers is not in the document]
Comment (berceanu) [reply unyt/__init__.py]: Done.

File: unyt/array.py [hunk 5]
```diff
@@ -2280,3 +2281,37 @@ def savetxt(
         newline="\n",
         comments=comments,
     )
+
+
+def allclose_units(actual, desired, rtol=1e-7, atol=0, **kwargs):
```
[the remark this answers is not in the document]
Comment (berceanu) [reply unyt/array.py]: Should it contain example usage and/or tests?

**Comment by berceanu** [comment]
> For the docstring, if you can please include an example. Look at the other docstrings in the file for templates.

Do you mean the docstring for `assert_allclose_units` or for `allclose_units`?

File: unyt/array.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply unyt/array.py]: Yup, the example and test are the same thing, the examples get executed during the tests and we make sure the output always stays the same.

File: unyt/array.py (hunk 2 above)
Comment (ngoldbaum) [reply unyt/array.py]: Something like this, I think:

```
with pytest.raises(UnitOperationError):
    allclose_units([1, 2, 3]*kg, [1, 2, 3]*kg, atol=3*meter)
```

(haven't actually tried running that, please double-check me)

**Comment by ngoldbaum** [comment]
I mean the new docstring you'll be adding for `allclose_units`. If `assert_allclose_units` doesn't have an example, it probably should as well :)

**Comment by berceanu** [comment]
All done.

**Comment by ngoldbaum** [comment]
Awesome, thanks so much for sticking with this and carrying it to completion. I hope this builds your confidence to contribute and I get to review more code from you :)

**Pull request merged**